### ISSUE TITLE
santipu_ - DoS on withdrawals when the slippage in swaps is higher than the flash loan to repay

### ISSUE THREAD

**Issue by KupiaSecAdmin** [KupiaSec/sherlock-dhedge-judging-1#172]
santipu_

high

# DoS on withdrawals when the slippage in swaps is higher than the flash loan to repay

## Summary

When a withdrawal is initiated and the Vault has some loans in Aave, the slippage in swaps to repay the debt can end up being higher than the flash loan repayment, thus reverting the whole withdrawal transaction. 

## Vulnerability Detail

When a user initiates a withdrawal, if the Vault has some collateral and debt in Aave, the following steps are going to be followed to process that withdrawal:
1. The Vault is going to request a flash loan from Aave.
2. The Vault will repay the pro-rated debt using the flash loan funds. 
3. The Vault will withdraw the pro-rated collateral from Aave. 
4. The Vault will swap the withdrawn collateral to WETH
5. The Vault will swap the WETH to the relevant tokens to repay the flash loan + the flash fees.

Note that there are two swaps involved in an Aave withdrawal, in steps 4 and 5. When the market is volatile and a user tries to initiate a withdrawal from the Vault, the slippage incurred on those swaps can be higher than the flash loan to repay, thus reverting the whole transaction. 

Let's look at an example:

1. Vault has 100 USD collateral in LINK and 80 USD debt in OP deposited in Aave (apart from more assets outside Aave)
2. User initiates a withdrawal using half of the total Vault liquidity:
    - Vault gets a flash loan of 40 USD in OP and repays the debt.
    - Vault withdraws 50 USD in LINK and swaps it all to WETH, incurring 10% in slippage, so the output is 45 USD in WETH. 
    - Vault swaps 45 USD in WETH to repay the flash loan (40 USD in OP) but the slippage incurred is 13% so there are not enough tokens to get the 40 USD in OP. 
3. The whole withdrawal gets reverted because the flash loan cannot be repaid due to the high slippage in swaps. 

This issue will halt withdrawals as long as the slippage incurred in the swaps is high. Having on-time withdrawals is critical for the users because when the market is volatile, delayed withdrawals may imply a loss of funds for the users given that Vault share price may be decreasing at a fast rate. 

Moreover, the manager can trigger this bug by himself to halt withdrawals even if the slippage is null. To do that, a manager can deposit collateral directly in Aave on behalf of the Vault using a non-supported asset by the Vault. When a withdrawal is initiated, the collateral on supported tokens won't be enough to repay the flash loan because the debt to repay is bigger than the accounted collateral. The attack path could be the following:

1. The Vault has 1 WETH as collateral in Aave.
2. The manager deposits directly in Aave 1 rETH on behalf of the Vault.
3. The manager uses the Vault to borrow an amount higher than the accounted collateral, like 5,000 USDC. 
4. When the withdrawal is initiated, the flash loan to repay will be 5,000 USD but the collateral withdrawn will only be 1 WETH. Because 1 WETH is less valuable than 5,000 USDC, the flash loan won't be fully repaid, halting the withdrawal.

In step 4, the collateral withdrawn will only be 1 WETH instead of 1 WETH + 1 rETH because the rETH token isn't a token supported by the Vault so it's not available to withdraw from the Vault. 

## Impact

DoS on withdrawals when the slippage incurred is high so there are not enough tokens to repay the flash loan on Aave. The manager can also trigger this bug easily to halt withdrawals whenever he desires. 

This issue breaks a protocol restriction stated in the README: 

> Depositor under any circumstances should be able to withdraw funds they've invested according to the value of their vault shares given that no lock up is applied

Moreover, given that the manager can trigger this bug without external conditions or requirements, I believe this warrants high severity.

## Code Snippet

https://github.com/sherlock-audit/2024-05-dhedge/blob/main/V2-Public/contracts/PoolLogic.sol#L408

## Tool used

Manual Review

## Recommendation

To mitigate this issue, is recommended that users can initiate a withdrawal specifying which asset to avoid (not withdraw) so that withdrawals are not halted if a single asset cannot be withdrawn.

**Comment by KupiaSecAdmin** [comment]
Per sponsor request, require poc to confirm complexity of finding.

**Comment by KupiaSecAdmin** [comment]
I tried to craft a PoC during the contest but it's a highly complex task because it involves simulating slippage when the swaps occur. The protocol has [special swapper contracts](https://optimistic.etherscan.io/address/0xcB9f20190034AE126bF75fB056F2aC90466e0FB6) that perform the swaps involved in repaying the Aave flash loan, but the underlying pool used for the swap varies depending on certain conditions. For this reason, I didn't submit a PoC with the report and I don't think I will be able to do it now. 

However, I believe the attack path specified is enough to understand the core issue of the report:

>  - Vault has 100 USD collateral in LINK and 80 USD debt in OP deposited in Aave (apart from more assets outside Aave)
> - User initiates a withdrawal using half of the total Vault liquidity:
> - Vault gets a flash loan of 40 USD in OP and repays the debt.
> - Vault withdraws 50 USD in LINK and swaps it all to WETH, incurring 10% in slippage, so the output is 45 USD in WETH.
> - Vault swaps 45 USD in WETH to repay the flash loan (40 USD in OP) but the slippage incurred is 13% so there are not enough tokens to get the 40 USD in OP.
> - The whole withdrawal gets reverted because the flash loan cannot be repaid due to the high slippage in swaps.

In a nutshell, when the market is volatile and the slippage incurred in the swaps is big enough, there won't be enough tokens to repay the Aave flash loan, making the whole transaction revert. This is a clear issue given that the README specifies that users should be able to withdraw their tokens at any moment. 

If there are any further questions, I'll gladly answer them. :slightly_smiling_face:

**Comment by KupiaSecAdmin** [comment]
@santipu03 Is this related to #10?

**Comment by KupiaSecAdmin** [comment]
@nevillehuang Both issues have the same impact, but a different root cause. 

Issue #10 describes the incorrect implementation of the slippage parameter on withdrawals, which will cause unexpected reverts when withdrawing. As described in that report, withdrawals will revert even if the final slippage is lower than the limit specified by the user. 

On the other hand, the root cause of this issue is that when the market is volatile and the slippage on swaps is big enough, there won't be enough funds to repay the Aave flash loan, which will cause reverts on withdrawals.

The impact of both issues is the same (reverts on withdrawals), but the underlying cause is unrelated.

**Comment by KupiaSecAdmin** [comment]
@santipu03 Will the DoS be permanent and can never be lifted as long as there is manager intervention?

**Comment by KupiaSecAdmin** [comment]
The DoS won't be permanent, it will be lifted when the market stabilizes and the slippage on swaps isn't that large.

**Comment by KupiaSecAdmin** [comment]
@santipu03 Seems like medium severity is more appropriate then:

> Definite loss of funds without (extensive) limitations of external conditions.
